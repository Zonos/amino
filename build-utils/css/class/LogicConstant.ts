import type { ConstantCustomizedComment } from 'build-utils/css/constants/logic/types/LogicConstantType';
import { capitalize } from 'build-utils/css/utils/capitalize';
import { formatTS } from 'build-utils/css/utils/formatTS';
import { readFileSync, writeFileSync } from 'fs';
import path from 'path';

type FileToParse = {
  /** @info file path to load (Need to be relative path - start from root of the project) */
  filePath: string;
};

type ParsedFile = FileToParse & {
  generateJsDocsFunction: ConstantCustomizedComment;
  hasJSDocsComment: boolean;
  keyValueParsed: Record<string, string> | null;
  rootFolder: string;
};

/**
 * Parse `ts` file from given path that has all the logic to generate key/value
 * and consume and generate constant
 */
export class LogicConstant {
  private parsedFile: ParsedFile;

  constructor() {
    const rootFolder = process.cwd();
    this.parsedFile = {
      filePath: '',
      generateJsDocsFunction: null,
      hasJSDocsComment: false,
      keyValueParsed: null,
      rootFolder,
    };
  }

  private printError(msg: string) {
    console.error(`File error '${__filename}': ${msg}`);
  }

  /**
   * Read and parse content from file to processable format
   * @param {FileToParse}
   */
  async parse({ filePath }: FileToParse): Promise<boolean> {
    this.parsedFile.filePath = filePath;
    if (filePath) {
      const parsedFileValid = this.checkValidParsedFile();
      if (!parsedFileValid) {
        this.printError(`Failed to parse file ${filePath}`);
      }
      return parsedFileValid;
    }
    this.printError('File path provided is invalid');
    return false;
  }

  private getFileName() {
    const { filePath } = this.parsedFile;
    return filePath.replace(/.+\/_?(.+)\.ts/gi, '$1');
  }

  /**
   * Check if file content satisfy naming rule for file before parsed or not
   */
  private async checkValidParsedFile(): Promise<boolean> {
    const { filePath, rootFolder } = this.parsedFile;
    /** Extract file name out of file path (without _) */
    const fileName = this.getFileName();
    const pascalFileName = capitalize(fileName);

    try {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const importResult: Record<string, any> = await import(
        `${rootFolder}/${filePath}`
      );

      /** Make sure function to generate content exist */
      const requiredGenerateContentFunction = `get${pascalFileName}ConstantKeyValuePairs`;
      if (!importResult[requiredGenerateContentFunction]) {
        throw Error(`${requiredGenerateContentFunction} function is required`);
      }

      this.parsedFile.keyValueParsed =
        importResult[requiredGenerateContentFunction]();

      /** Make sure function to generate JSDocs exist */
      const requiredGenerateCommentFunction = `get${pascalFileName}ConstantCustomizedComment`;

      if (!importResult[requiredGenerateCommentFunction]) {
        throw Error(`${requiredGenerateCommentFunction} function is required`);
      }
      this.parsedFile.generateJsDocsFunction =
        importResult[requiredGenerateCommentFunction];

      /** Make sure `hasJSDocsComment` exist */
      const requiredHasJSDocsFlag = `hasJSDocsComment`;
      if (importResult[requiredHasJSDocsFlag] === undefined) {
        throw Error(`'${requiredHasJSDocsFlag}' constant is required.`);
      }
      this.parsedFile.hasJSDocsComment = importResult[requiredHasJSDocsFlag];
    } catch (ex) {
      if (ex instanceof Error) {
        this.printError(ex.message);
        return false;
      }
    }
    return true;
  }

  generateConstantContent() {
    const { generateJsDocsFunction, hasJSDocsComment, keyValueParsed } =
      this.parsedFile;
    if (keyValueParsed) {
      const contents = Object.entries(keyValueParsed).map(([key, value]) =>
        [
          /** Add JSDocsComment by adding content get function `generateJsDocsFunction` on each loop */
          hasJSDocsComment && generateJsDocsFunction
            ? `/** ${generateJsDocsFunction({ key, value })} */`
            : '',
          `'${key}': '${value}',`,
        ]
          .filter(Boolean)
          .join('\n'),
      );
      return contents.join('\n');
    }
    return null;
  }

  async generateConstantFile() {
    const { rootFolder } = this.parsedFile;
    const constantContent = this.generateConstantContent();
    const fileName = this.getFileName();
    if (constantContent) {
      const fileContent = `
      export const ${fileName} = {
        ${constantContent}
      } as const;
    `;
      const formatedContent = await formatTS(fileContent);
      writeFileSync(
        `${rootFolder}/build-utils/css/constants/generated/_${fileName}.ts`,
        formatedContent,
      );
    }
  }

  /** Checking camel case for a text */
  static camelCaseChecking(text: string) {
    return !!text && /^[a-z]/.test(text) && !/[\W_]/.test(text);
  }

  /**
   *
   * @param fileName file name that would be generated.
   *
   * **Note**:
   * - Need to be camelCase
   * - No extension
   * @return
   * - `null` when fileName is invalid
   * - `string` when fileName is valid
   */
  static async _generateLogicConstFileContent(
    fileName: string,
  ): Promise<string> {
    if (!this.camelCaseChecking(fileName)) {
      throw Error(
        'File name needs to be in valid camelCase format with no extension and no underscore.',
      );
    }
    const rootFolder = process.cwd();
    /** Read the template content */
    const templateContent = readFileSync(
      path.resolve(
        `${rootFolder}/build-utils/css/class/template/logicConstant.tpl`,
      ),
      { encoding: 'utf-8' },
    );
    const newTemplateContent = templateContent
      .replace(/{{FileName}}/g, fileName)
      .replace(/{{CapitalizedFileName}}/g, capitalize(fileName));

    return formatTS(newTemplateContent);
  }

  /**
   *
   * @param fileName file name that would be generated.
   * File name:
   * - Need to be camelCase
   * - Have no extension
   * @note
   * - Generated file will be located at `build-utils/css/constants/logic` with file name `_${fileName}.ts`.
   * - It will do nothing if fileName is in wrong format has nothing
   */
  static async generateLogicConstantFile(fileName: string) {
    const fileContent = await this._generateLogicConstFileContent(fileName);
    const rootFolder = process.cwd();
    writeFileSync(
      `${rootFolder}/build-utils/css/constants/logic/_${fileName}.ts`,
      fileContent,
    );
  }

  /**
   *
   * @param content : file content that need to converted;
   * @returns
   */
  static async transformImportedConstant(content: string) {
    const matchedImportLines = content.matchAll(
      /import {\s+(\w+)\s+} from '(.+)';/gm,
    );
    const matchedArr = Array.from(matchedImportLines);

    /** Key value pair of extracted constants */
    const extractedConstants = matchedArr.map(([, variable, filePath]) => {
      const rootFolder = process.cwd();
      /** Read generated file content */
      const fileContent = readFileSync(
        path.resolve(`${rootFolder}/${filePath}.ts`),
        { encoding: 'utf-8' },
      );
      /** Extract only constant content */
      const matchAll = fileContent.matchAll(
        /(?<=.+ = {\n +)((\n|.)+)(?=}.+)/gm,
      );
      const matched = Array.from(matchAll);
      const result = matched
        .map(([extractedVariable]) => extractedVariable)
        .join('');
      return {
        data: result,
        key: variable,
      };
    });

    const result = content
      /** Clean up all imports */
      .replace(/import {\s+(\w+)\s+} from '(.+)';/gm, '')
      /** Replace spread constant in theme constant with extracted constant from generated constant file */
      .replace(
        /\.\.\.(\w+),/gm,
        (_g0, variable) =>
          extractedConstants.find(item => item.key === variable)?.data || _g0,
      );
    return formatTS(result);
  }
}
