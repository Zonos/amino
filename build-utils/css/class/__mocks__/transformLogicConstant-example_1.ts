export const example1 = {
  key1: 'value1',
  key2: 'value2',
} as const;
