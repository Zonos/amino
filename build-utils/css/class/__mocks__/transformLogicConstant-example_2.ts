export const example2 = {
  keyExample21: 'value Example2 1',
  keyExample22: 'value Example2 2',
} as const;
