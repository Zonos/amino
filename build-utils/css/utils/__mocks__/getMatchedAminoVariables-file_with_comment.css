:root {
  --amino-transition: all 0.2s ease-in-out 0s;
  --amino-type-scale-base: 16px;
  --amino-font-size-3xl: 2.125rem; /* 34px */
  --amino-font-size-2xl: 1.75rem;  /* 28px */
  --amino-font-size-xl: 1.375rem;  /* 22px */
  --amino-font-size-l: 1.125rem;   /* 18px */
  --amino-font-size-base: .875rem;    /* 14px */
  --amino-font-size-s: .75rem;    /* 12px */
  --amino-font-size-xs: .625rem;   /* 10px */
  --amino-line-height-3xl: 118%;   /* 40px */
  --amino-line-height-2xl: 119%;   /* 32px */
  --amino-line-height-xl: 127%;    /* 28px */
  --amino-line-height-l: 133%;     /* 16px */
  --amino-line-height-base: 114%;  /* 16px */
  --amino-line-height-s: 133%;     /* 16px */
  --amino-line-height-xs: 120%;    /* 12px */
  /* legacy text variables */
  --amino-text-xxl: var(--amino-font-size-3xl);
  --amino-text-xl: var(--amino-font-size-2xl);
  --amino-text-lg: var(--amino-font-size-xl);
  --amino-text-md: var(--amino-font-size-l);
  --amino-text-base: var(--amino-font-size-base);
  --amino-text-sm: var(--amino-font-size-s);
  --amino-text-xs: var(--amino-font-size-xs);
  --amino-space: 1.5rem; /* 24px */
  --amino-space-negative: -1.5rem; /* -24px */
  --amino-space-half: 1rem; /* 16px */
  --amino-space-half-negative: -1rem; /* -16px */
  --amino-space-quarter: 0.5rem; /* 8px */
  --amino-space-quarter-negative: -0.5rem; /* -8px */
  --amino-space-double: 2.5rem; /* 40px */
  --amino-space-double-negative: -2.5rem; /* -40px */
  --amino-radius-sm: 0.25rem; /* 4px */
  --amino-radius: 0.375rem; /* 6px */
  --amino-radius-lg: 0.5rem; /* 8px */
  --amino-radius-xl: 0.75rem; /* 12px */

  /* GRAY PALETTE */
  --amino-gray-l80: #f5f5f6;
  --amino-gray-l60: #eaeaec;
  --amino-gray-l40: #d6d6d9;
  --amino-gray-l20: #b7b7bd;
  --amino-gray-base: #9898a0;
  --amino-gray-d20: #7a7a80;
  --amino-gray-d40: #5b5b60;
  --amino-gray-d60: #3d3d40;
  --amino-gray-d80: #232326;
  --amino-gray-50: #f9fafb;
  --amino-gray-100: var(--amino-gray-l80);
  --amino-gray-200: var(--amino-gray-l60);
  --amino-gray-300: var(--amino-gray-l40);
  --amino-gray-400: var(--amino-gray-l20);
  --amino-gray-500: var(--amino-gray-base);
  --amino-gray-600: var(--amino-gray-d20);
  --amino-gray-700: var(--amino-gray-d40);
  --amino-gray-800: var(--amino-gray-d60);
  --amino-gray-900: var(--amino-gray-d80);
  /* BLUE PALETTE */
  --amino-blue-l80: #e9ebff;
  --amino-blue-l60: #a7afff;
  --amino-blue-l40: #7b86ff;
  --amino-blue-l20: #4f5eff;
  --amino-blue-base: #2337ff;
  --amino-blue-d20: #1c2aca;
  --amino-blue-d40: #152099;
  --amino-blue-d60: #0f1666;
  --amino-blue-d80: #1d233b;
  --amino-blue-alpha: #2190ff80;
  --amino-blue-50: #ebf5ff;
  --amino-blue-100: var(--amino-blue-l80);
  --amino-blue-200: var(--amino-blue-l60);
  --amino-blue-300: var(--amino-blue-l40);
  --amino-blue-400: var(--amino-blue-l20);
  --amino-blue-500: var(--amino-blue-base);
  --amino-blue-600: var(--amino-blue-d20);
  --amino-blue-700: var(--amino-blue-d40);
  --amino-blue-800: var(--amino-blue-d60);
  --amino-blue-900: var(--amino-blue-d80);
  /* CYAN PALETTE */
  --amino-cyan-l80: #e9fcff;
  --amino-cyan-l60: #a5edf6;
  --amino-cyan-l40: #77e3f2;
  --amino-cyan-l20: #4adaed;
  --amino-cyan-base: #1dd1e9;
  --amino-cyan-d20: #17a7ba;
  --amino-cyan-d40: #117d8c;
  --amino-cyan-d60: #0c545d;
  --amino-cyan-d80: #1d363b;
  --amino-cyan-alpha: #00dfe380;
  --amino-cyan-100: var(--amino-cyan-l80);
  --amino-cyan-200: var(--amino-cyan-l60);
  --amino-cyan-300: var(--amino-cyan-l40);
  --amino-cyan-400: var(--amino-cyan-l20);
  --amino-cyan-500: var(--amino-cyan-base);
  --amino-cyan-600: var(--amino-cyan-d20);
  --amino-cyan-700: var(--amino-cyan-d40);
  --amino-cyan-800: var(--amino-cyan-d60);
  --amino-cyan-900: var(--amino-cyan-d80);
  /* RED PALETTE */
  --amino-red-l80: #fde5eb;
  --amino-red-l60: #f899b1;
  --amino-red-l40: #f5668a;
  --amino-red-l20: #f13363;
  --amino-red-base: #ee003c;
  --amino-red-d20: #be0030;
  --amino-red-d40: #8f0024;
  --amino-red-d60: #5f0018;
  --amino-red-d80: #371822;
  --amino-red-alpha: #e0451a80;
  --amino-red-50: #fef2f2;
  --amino-red-100: var(--amino-red-l80);
  --amino-red-200: var(--amino-red-l60);
  --amino-red-300: var(--amino-red-l40);
  --amino-red-400: var(--amino-red-l20);
  --amino-red-500: var(--amino-red-base);
  --amino-red-600: var(--amino-red-d20);
  --amino-red-700: var(--amino-red-d40);
  --amino-red-800: var(--amino-red-d60);
  --amino-red-900: var(--amino-red-d80);
  /* ORANGE PALETTE */
  --amino-orange-l80: #fff0e5;
  --amino-orange-l60: #ffc499;
  --amino-orange-l40: #ffa666;
  --amino-orange-l20: #ff8933;
  --amino-orange-base: #ff6b00;
  --amino-orange-d20: #cc5600;
  --amino-orange-d40: #994000;
  --amino-orange-d60: #662b00;
  --amino-orange-d80: #39271c;
  --amino-orange-alpha: #ffa51e80;
  --amino-orange-100: var(--amino-orange-l80);
  --amino-orange-200: var(--amino-orange-l60);
  --amino-orange-300: var(--amino-orange-l40);
  --amino-orange-400: var(--amino-orange-l20);
  --amino-orange-500: var(--amino-orange-base);
  --amino-orange-600: var(--amino-orange-d20);
  --amino-orange-700: var(--amino-orange-d40);
  --amino-orange-800: var(--amino-orange-d60);
  --amino-orange-900: var(--amino-orange-d80);
  /* GREEN PALETTE */
  --amino-green-l80: #e5f9e5;
  --amino-green-l60: #99e799;
  --amino-green-l40: #66db66;
  --amino-green-l20: #33cf33;
  --amino-green-base: #00c300;
  --amino-green-d20: #009c00;
  --amino-green-d40: #007500;
  --amino-green-d60: #004e00;
  --amino-green-d80: #17311a;
  --amino-green-alpha: #ecfdf580;
  --amino-green-50: #ecfdf5;
  --amino-green-100: var(--amino-green-l80);
  --amino-green-200: var(--amino-green-l60);
  --amino-green-300: var(--amino-green-l40);
  --amino-green-400: var(--amino-green-l20);
  --amino-green-500: var(--amino-green-base);
  --amino-green-600: var(--amino-green-d20);
  --amino-green-700: var(--amino-green-d40);
  --amino-green-800: var(--amino-green-d60);
  --amino-green-900: var(--amino-green-d80);
  /* PURPLE */
  --amino-purple-l80: #f5e5ff;
  --amino-purple-l60: #d599ff;
  --amino-purple-l40: #c066ff;
  --amino-purple-l20: #ab33ff;
  --amino-purple-base: #9600ff;
  --amino-purple-d20: #7800cc;
  --amino-purple-d40: #5a0099;
  --amino-purple-d60: #3c0066;
  --amino-purple-d80: #2c183b;
  --amino-purple-alpha: #f5f3ff80;
  --amino-purple-50: #f5f3ff;
  --amino-purple-100: var(--amino-purple-l80);
  --amino-purple-200: var(--amino-purple-l60);
  --amino-purple-300: var(--amino-purple-l40);
  --amino-purple-400: var(--amino-purple-l20);
  --amino-purple-500: var(--amino-purple-base);
  --amino-purple-600: var(--amino-purple-d20);
  --amino-purple-700: var(--amino-purple-d40);
  --amino-purple-800: var(--amino-purple-d60);
  --amino-purple-900: var(--amino-purple-d80);
  /* YELLOW PALLETE */
  --amino-yellow-l80: #fff9e5;
  --amino-yellow-l60: #ffe999;
  --amino-yellow-l40: #ffdd66;
  --amino-yellow-l20: #ffd233;
  --amino-yellow-base: #ffc700;
  --amino-yellow-d20: #cc9f00;
  --amino-yellow-d40: #997700;
  --amino-yellow-d60: #665000;
  --amino-yellow-d80: #39321c;
  --amino-yellow-100: var(--amino-yellow-l80);
  --amino-yellow-200: var(--amino-yellow-l60);
  --amino-yellow-300: var(--amino-yellow-l40);
  --amino-yellow-400: var(--amino-yellow-l20);
  --amino-yellow-500: var(--amino-yellow-base);
  --amino-yellow-600: var(--amino-yellow-d20);
  --amino-yellow-700: var(--amino-yellow-d40);
  --amino-yellow-800: var(--amino-yellow-d60);
  --amino-yellow-900: var(--amino-yellow-d80);
  --amino-amber-alpha: #fffbeb80;
  --amino-amber-50: #fffbeb;
  --amino-amber-100: var(--amino-yellow-l80);
  --amino-amber-200: var(--amino-yellow-l60);
  --amino-amber-300: var(--amino-yellow-l40);
  --amino-amber-400: var(--amino-yellow-l20);
  --amino-amber-500: var(--amino-yellow-base);
  --amino-amber-600: var(--amino-yellow-d20);
  --amino-amber-700: var(--amino-yellow-d40);
  --amino-amber-800: var(--amino-yellow-d60);
  --amino-amber-900: var(--amino-yellow-d80);
  --amino-danger: var(--amino-red-500);
  --amino-danger-dark: var(--amino-red-600);
  --amino-error: var(--amino-red-500);
  --amino-glow-error: 0 0 0 1px var(--amino-red-500);
  --amino-glow-blue: 0 0 0 1px var(--amino-primary);
  --amino-glow-cyan: 0 0 0 1px var(--amino-cyan-alpha);
  --amino-glow-red: 0 0 0 1px var(--amino-red-alpha);
  --amino-glow-orange: 0 0 0 1px var(--amino-orange-alpha);
  --amino-glow-green: 0 0 0 1px var(--amino-green-alpha);
  --amino-glow-purple: 0 0 0 1px var(--amino-purple-alpha);
  --amino-primary: var(--amino-blue-500);
  --amino-primary-light: var(--amino-blue-400);
  --amino-primary-dark: var(--amino-blue-700);
  --amino-success: var(--amino-green-200);
  --amino-success-dark: var(--amino-green-600);
  --amino-warning: var(--amino-amber-100);
  --amino-warning-dark: var(--amino-amber-600);
  --amino-page-background: white;
  --amino-border-color: var(--amino-gray-200);
  --amino-hover-color: var(--amino-gray-50);
  --amino-surface-color: white;
  --amino-surface-color-secondary: var(--amino-gray-50);
  --amino-text-light: white;
  --amino-text-color: var(--amino-gray-900);
  --amino-text-dark: var(--amino-text-color);
  --amino-input-background: white;
  --amino-size-xl: 56px;
  --amino-size-lg: 48px;
  --amino-size-md: 40px;
  --amino-size-sm: 32px;
  --amino-appbar-height: 55px;
  --amino-sidebar-width: 264px;
  --amino-sidebar-color: white;
  --amino-header-color: white;
  --amino-backdrop-color: var(--amino-gray-900);
  --amino-elevation-0: 0;
  --amino-elevation-100: 100;
  --amino-elevation-200: 200;
  --amino-elevation-300: 300;
  --amino-elevation-400: 400;
  --amino-elevation-500: 500;
  --amino-dialog-elevation: var(--amino-elevation-400);
  --amino-toast-elevation: var(--amino-elevation-500);
  --amino-appbar-elevation: var(--amino-elevation-500);
  --amino-font-sans: 'Inter', -apple-system, BlinkMacSystemFont, 'Segoe UI',
    'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans',
    'Helvetica Neue', sans-serif;
  --amino-font-mono: Operator Mono, MonoLisa, Dank Mono, Menlo, Monaco,
    Lucida Console, Liberation Mono, DejaVu Sans Mono, Bitstream Vera Sans Mono,
    Courier New, monospace;
  /* SHADOW SETTING */
  --amino-v3-shadow-inset: inset 0px 2px 4px rgba(0, 0, 0, 0.06);
  --amino-v3-shadow-base: 0px 1px 2px rgba(0, 0, 0, 0.06),
    0px 1px 3px rgba(0, 0, 0, 0.1);
  --amino-v3-shadow-medium: 0px 2px 4px rgba(0, 0, 0, 0.06),
    0px 4px 6px rgba(0, 0, 0, 0.1);
  --amino-v3-shadow-large: 0px 10px 15px rgba(0, 0, 0, 0.1),
    0px 4px 6px rgba(0, 0, 0, 0.05);
  --amino-v3-shadow-xl: 0px 20px 25px rgba(0, 0, 0, 0.1),
    0px 10px 10px rgba(0, 0, 0, 0.04);
  --amino-v3-shadow-xxl: 0px 32px 50px rgba(0, 0, 0, 0.24);
  --amino-shadow-inset: var(--amino-v3-shadow-inset);
  --amino-shadow-small: var(--amino-v3-shadow-base);
  --amino-shadow-base: var(--amino-v3-shadow-medium);
  --amino-shadow-medium: var(--amino-v3-shadow-large);
  --amino-shadow-large: var(--amino-v3-shadow-xl);
  --amino-shadow-larger: var(--amino-v3-shadow-xxl);
  /* helpers */
  --amino-border: 1px solid var(--amino-border-color);
  --amino-border-blue: 1px solid var(--amino-blue-400);
  --amino-border-red: 1px solid var(--amino-red-300);
  --amino-border-cyan: 1px solid var(--amino-cyan-300);
  --amino-border-orange: 1px solid var(--amino-orange-300);
  --amino-border-green: 1px solid var(--amino-green-300);
  --amino-border-purple: 1px solid var(--amino-purple-300);
  --amino-border-amber: 1px solid var(--amino-amber-300);
  --amino-border-transparent: 1px solid transparent;
  /* NOTE TO ADD: easing functions and timings */
  /* NOTE TO ADD: test contrast ratios */
  /* NOTE TO ADD: more accessible hover color */
}
