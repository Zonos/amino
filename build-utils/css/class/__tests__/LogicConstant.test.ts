import { existsSync, unlinkSync } from 'fs';

import { LogicConstant } from '../LogicConstant';

describe('Make sure transformLogicConstant in LogicConstant work properly', () => {
  type UnitTest = {
    case: string;
    content: string;
  };
  const testCases: UnitTest[] = [
    {
      case: 'Has 1 import',
      content: `
import { example1 } from '../class/__mocks__/transformLogicConstant-example_1';

export const test = {
  ...example1,
} as const;
            `,
    },
    {
      case: 'Has 2 imports',
      content: `
import { example1 } from '../class/__mocks__/transformLogicConstant-example_1';
import { example2 } from '../class/__mocks__/transformLogicConstant-example_2';

export const test = {
  ...example1,

  /** Example 2 */
  ...example2,
} as const;
            `,
    },
  ];
  test.each(testCases)(`Case: '$case'`, async ({ case: _case, content }) => {
    const result = await LogicConstant.transformImportedConstant(content);
    expect(result).toMatchSnapshot(_case);
  });
});

describe('Make sure `camelCaseChecking` in LogicConstant work properly', () => {
  type UnitTest = {
    text: string;
    expected: boolean;
  };
  const testCases: UnitTest[] = [
    {
      text: '',
      expected: false,
    },
    {
      text: 'PasCalCase',
      expected: false,
    },
    {
      text: 'camelCase',
      expected: true,
    },
    {
      text: 'camelcase',
      expected: true,
    },
    {
      text: 'camelCaseWithNumber123',
      expected: true,
    },
    {
      text: 'just_SnakeCase',
      expected: false,
    },
    {
      text: 'Mix_SnakeCase_And_PasCal',
      expected: false,
    },
    {
      text: '123camelCaseWithNumber',
      expected: false,
    },
    {
      text: 'camelCaseWithNumber123',
      expected: true,
    },
    {
      text: 'sp3c1alCh@r@ct3rCame',
      expected: false,
    },
  ];
  test.each(testCases)(
    `text: $text - expected: $expected`,
    async ({ text, expected }) => {
      const result = LogicConstant.camelCaseChecking(text);
      expect(result).toBe(expected);
    }
  );
});

describe('Make sure `generateLogicConstFileContent` in LogicConstant work properly', () => {
  test(`Generated template content match snapshot`, async () => {
    const result = await LogicConstant._generateLogicConstFileContent('space');
    expect(result).toMatchSnapshot();
  });

  test(`Passing empty file name to the function`, async () => {
    try {
      await LogicConstant._generateLogicConstFileContent('');
    } catch (err) {
      if (err instanceof Error) {
        expect(err.message).toBe(
          'File name needs to be in valid camelCase format with no extension and no underscore.'
        );
      }
    }
  });
});

describe('Make sure `generateLogicConstFile` in LogicConstant work properly', () => {
  const rootFolder = process.cwd();
  const fileName = 'testSpaceSnapshot';
  const filePath = `${rootFolder}/build-utils/css/constants/logics/_${fileName}.ts`;

  afterEach(() => {
    // clean up the file after generating it
    if (existsSync(filePath)) {
      unlinkSync(filePath);
    }
  });
  test(`New file generated at 'build-utils/css/logics'`, async () => {
    await LogicConstant.generateLogicConstantFile(fileName);
    const exist = existsSync(filePath);
    expect(exist).toBe(true);
  });
  test(`Passing empty file name to the function`, async () => {
    try {
      await LogicConstant.generateLogicConstantFile('');
    } catch (err) {
      if (err instanceof Error) {
        expect(err.message).toBe(
          'File name needs to be in valid camelCase format with no extension and no underscore.'
        );
      }
    }
  });
  test(`Passing invalid file name to the function`, async () => {
    try {
      await LogicConstant.generateLogicConstantFile('__invalidFileName');
    } catch (err) {
      if (err instanceof Error) {
        expect(err.message).toBe(
          'File name needs to be in valid camelCase format with no extension and no underscore.'
        );
      }
    }
  });
});
