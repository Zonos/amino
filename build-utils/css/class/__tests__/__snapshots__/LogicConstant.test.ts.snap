// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make sure \`generateLogicConstFileContent\` in LogicConstant work properly Generated template content match snapshot 1`] = `
"import {
  ConstantCustomizedComment,
  ConstantKeyValuePairsType,
} from './types/LogicConstantType';

/** Flag to determine to apply jdocs generation rule from getSpaceConstantCustomizedComment function below */
export const hasJSDocsComment: boolean = true;

/**
 * Get jsdocs comment that need to be injected to each space constant with customized comment
 *
 * **NOTE**: only enable when \`hasJSDocsComment\` is on
 * */
export const getSpaceConstantCustomizedComment: ConstantCustomizedComment = ({
  key,
  value,
}) => {
  /** Put logic here to generate jsdocs string for each line in constant */
  return \`\${key}: \${value}\`;
};

/**
 * Return key/value pairs after applying logic
 */
export const getSpaceConstantKeyValuePairs: ConstantKeyValuePairsType = () => {
  const contentArr: Record<string, string> = {};

  /** Put logic here to generate constant key value pairs */

  return contentArr;
};
"
`;

exports[`Make sure transformLogicConstant in LogicConstant work properly Case: 'Has 1 import': Has 1 import 1`] = `
"export const test = {
  key1: 'value1',
  key2: 'value2',
} as const;
"
`;

exports[`Make sure transformLogicConstant in LogicConstant work properly Case: 'Has 2 imports': Has 2 imports 1`] = `
"export const test = {
  key1: 'value1',
  key2: 'value2',

  /** Example 2 */
  keyExample21: 'value Example2 1',
  keyExample22: 'value Example2 2',
} as const;
"
`;
